### PR TITLE
Expose SNP Attestation validation in TS

### PR DESCRIPTION
**Pull request by DomAyre**
Todo:
- [x]  Test with non static Attestation report
- [x] Compile SNP separately from openenclave
- [x] Do more validation in modules_test

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -448,7 +448,10 @@ set(CCF_NETWORK_TEST_ARGS
 )
 
 if(COMPILE_TARGET STREQUAL "sgx")
-  add_enclave_library(js_openenclave.enclave ${CCF_DIR}/src/js/openenclave.cpp)
+  add_enclave_library(
+    js_openenclave.enclave ${CCF_DIR}/src/js/openenclave.cpp
```
Comment (achamayou) [inline CMakeLists.txt:452]: Best to either split as you propose, or rename this to js_attestation_verification.

File: js/ccf-app/src/global.ts [hunk 2]
```diff
@@ -739,3 +739,19 @@ export interface OpenEnclave {
     endorsements?: ArrayBuffer,
   ): EvidenceClaims;
 }
+
+export interface SnpAttestationResult {
+  measurement: string;
```
Comment (achamayou) [inline js/ccf-app/src/global.ts:744]: These should be exposed as ArrayBuffers, they're not necessarily legal JS strings, so they'd need to be base64'd or hex'ed, and the application code can do that if it wants.

File: tests/js-modules/modules.py [hunk 3]
```diff
@@ -948,6 +947,20 @@ def test_npm_app(network, args):
...
+            LOG.info("SNP: Test verifySnpAttestation")
+            r = c.post(
+                "/app/verifySnpAttestation",
+                {
+                    "evidence": primary_quote_info["raw"],
+                    "endorsements": primary_quote_info["endorsements"],
+                    "endorsed_tcb": primary_quote_info["endorsed_tcb"],
+                },
+            )
+            assert r.status_code == http.HTTPStatus.OK, r.status_code
```
Comment (achamayou) [inline tests/js-modules/modules.py:960]: It would be good to test this a bit more: evidence too big, too short, endorsements that don't match, endorsed_tcb that doesn't match etc.

File: js/ccf-app/src/global.ts (hunk 2 above)
Comment (achamayou) [reply js/ccf-app/src/global.ts:744]: Plus it'd be consistent with what we expose the claims with the OE version already.

File: CMakeLists.txt (hunk 1 above)
Comment (jumaffre) [reply CMakeLists.txt:452]: I think splitting is better as verifying SGX quotes requires `open-enclave-hostverify` and we may want to build CCF with no dependency on OE whatsoever, but which can still verify SNP quotes.

File: tests/js-modules/modules.py (hunk 3 above)
Comment (DomAyre) [reply tests/js-modules/modules.py:960]: https://github.com/microsoft/CCF/pull/5653/commits/ff9aefa7538755b0d1fa8de035f490d39ab90523

File: src/apps/js_generic/js_generic.cpp [hunk 4]
```diff
@@ -15,7 +16,7 @@ namespace ccfapp
   std::vector<ccf::js::FFIPlugin> get_js_plugins()
   {
 #ifdef SGX_ATTESTATION_VERIFICATION
-    return {ccf::js::openenclave_plugin};
+    return {ccf::js::openenclave_plugin, ccf::js::snp_attestation_plugin};
```
Comment (jumaffre) [inline src/apps/js_generic/js_generic.cpp:19]: I think `ccf::js::snp_attestation_plugin` should be set even if SGX_ATTESTATION_VERIFICATION` is false.

File: tests/npm-app/src/endpoints/snp_attestation.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+export interface TcbVersion {
+  boot_loader: number;
+  tee: number;
+  snp: number;
+  microcode: number;
+}
+
+interface SnpAttestationResult {
```
Comment (achamayou) [inline tests/npm-app/src/endpoints/snp_attestation.ts:25]: why is this duplicated and not imported from ccf-app?

File: tests/js-modules/modules.py [hunk 6]
```diff
@@ -948,6 +947,99 @@ def test_npm_app(network, args):
...
+            # Test corrupted endorsements
+            r = c.post(
+                "/app/verifySnpAttestation",
+                {
+                    "evidence": primary_quote_info["raw"],
+                    "endorsements": corrupt_value(primary_quote_info["endorsements"]),
+                    "endorsed_tcb": primary_quote_info["endorsed_tcb"],
+                },
+            )
+            assert r.status_code != http.HTTPStatus.OK, r.status_code
```
Comment (achamayou) [inline tests/js-modules/modules.py:1040]: This is not good practice, if the code throws and returns a 500, this may pass for the wrong reason. We should check for the expected error code here (4xx probably?)

File: src/apps/js_generic/js_generic.cpp (hunk 4 above)
Comment (achamayou) [reply src/apps/js_generic/js_generic.cpp:19]: We don't need a gating flag for the snp attestation, since it does not drag in external deps.

File: src/js/snp_attestation.cpp [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+#include <algorithm>
+#include <quickjs/quickjs.h>
+#include <regex>
+#include <vector>
+
+namespace ccf::js
+{
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wc99-extensions"
```
Comment (jumaffre) [inline src/js/snp_attestation.cpp:18]: Why is this required?

**Review by achamayou — APPROVED** [review]

File: tests/npm-app/src/endpoints/snp_attestation.ts (hunk 5 above)
Comment (DomAyre) [reply tests/npm-app/src/endpoints/snp_attestation.ts:25]: In ccf-app, the values are array buffers, but here we convert them into hex strings so those types have changed, its the same pattern we use for the OE bindings

File: src/js/snp_attestation.cpp (hunk 7 above)
Comment (DomAyre) [reply src/js/snp_attestation.cpp]: https://github.com/microsoft/CCF/pull/5653/commits/b316adc7febba6ca788f11dc4b1771c6c3efde1f

File: tests/js-modules/modules.py (hunk 6 above)
Comment (DomAyre) [reply tests/js-modules/modules.py]: https://github.com/microsoft/CCF/pull/5653/commits/f5c4991f3d961435a114422a21a47d318c9a8be7

**Comment by achamayou** [comment]
@DomAyre could you add an entry to the changelog please, just describing the feature and the name of the new API? I think I've worked out why the SNP job did not trigger automatically on this PR, it should now.

File: CHANGELOG.md [hunk 8]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Updated `nghttp2` from `1.51.0` to `1.55.1`.
 - Converted SNP attestation UVM endorsements from integer to arbitrary string.
 - Updated Intel SGX PSW from 2.17 to 2.20 (#5616)
+- Added `/app/verifySnpAttestation` endpoint which verifies SNP Attestations. (#5653)
```
Comment (achamayou) [inline CHANGELOG.md:18]: That's in a test app, it's not the public API. This should instead document the introduction of ccf.SnpAttestation.verifySnpAttestation() as a public API for TypeScript apps.

File: src/js/snp_attestation.cpp [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    pal::PlatformAttestationMeasurement measurement = {};
+    pal::PlatformAttestationReportData report_data = {};
+    UVMEndorsements parsed_uvm_endorsements;
+
+    try
+    {
+      pal::verify_snp_attestation_report(quote_info, measurement, report_data);
+      parsed_uvm_endorsements =
```
Comment (achamayou) [inline src/js/snp_attestation.cpp:90]: This mandates uvm endorsements in the current format, and does not allow the caller to pass a trusted platform measurement instead, or to validate it against another form of endorsement.

Comment (achamayou) [reply src/js/snp_attestation.cpp:90]: Confirmed with Kapil that manual checks need to be supported, so this step needs to be made optional when no UVM endorsements are passed, and the caller will then check `measurement` separately.

Comment (DomAyre) [reply src/js/snp_attestation.cpp]: https://github.com/microsoft/CCF/pull/5653/commits/b317000c0b47e8fbe2c61cb53ac3d11051a776b8

File: CHANGELOG.md (hunk 8 above)
Comment (DomAyre) [reply CHANGELOG.md]: https://github.com/microsoft/CCF/pull/5653/commits/13bf121765a263ba2e1b4eede7eca5b3a8e4c4ac
